File: .leetcode/209.长度最小的子数组.cpp
/*
 * @lc app=leetcode.cn id=209 lang=cpp
 *
 * [209] 长度最小的子数组
 */
#include<bits/stdc++.h>
using namespace std;

// @lc code=start
class Solution {
public:
    int minSubArrayLen(int target, vector<int>& nums) {
        int left = -1,right = 0;
        int ans = INT_MAX;
        int tmp = 0;
        int all = 0;
        while(right < nums.size()){
            all += nums[right];
            tmp += nums[right];
            if(tmp >= target){
                while(left < right && tmp - nums[left + 1] >= target){
                    left++;
                    tmp -= nums[left];
                }
                ans = min(ans,right - left);
            }
            right++;
        }
        if(all < target)return 0;
        return ans;
    }
};
// @lc code=end

File: .leetcode/209.长度最小的子数组.java
/*
 * @lc app=leetcode.cn id=209 lang=java
 *
 * [209] 长度最小的子数组
 */

// @lc code=start
class Solution {
    public int minSubArrayLen(int target, int[] nums) {
        int ans = Integer.MAX_VALUE;
        int n = nums.length;
        int temp = 0,left = 0;
        for(int right = 0;right < n;right++){
            temp += nums[right];
            while(temp >= target && left <= right){
                ans = Math.min(ans, right - left + 1);
                temp -= nums[left++];
            }
        }
        return ans == Integer.MAX_VALUE?0:ans;
    }
}
// @lc code=end

